wordpress.org/public_html/wp-content/plugins/plugin-directory/cli/class-block-plugin-checker.php

```diff
@@ -0,0 +1,241 @@
+<?php
+namespace WordPressdotorg\Plugin_Directory\CLI;
+
+use WordPressdotorg\Plugin_Directory\Readme\Parser;
+use WordPressdotorg\Plugin_Directory\Tools\Filesystem;
+
+/**
+ * A class that can examine a plugin, and evaluate and return status info that would be useful for a validator or similar tool.
+ *
+ * Note: I've written this as one class for convenience, but as it's evolved I think it would make sense to split it in two: one for collecting data
+ * (particularly the prepare*() and find*() functions), and a separate class for the checks.
+ *
+ * @package WordPressdotorg\Plugin_Directory\CLI
+ */
+class Block_Plugin_Checker {
+
+	protected $path_to_plugin = null;
+	protected $check_methods = array();
+	protected $results = array();
+
+	protected $slug = null;
+	protected $readme_path = null;
+	protected $readme = null;
+	protected $headers = null;
+	protected $blocks = null;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param string $slug The plugin slug, if known. Optional.
+	 */
+	public function __construct( $slug = null ) {
+		$this->slug = $slug;
+	}
+
+	/**
+	 * Check a plugin that has been unzipped or exported to a local filesystem.
+	 *
+	 * @param string $path_to_plugin Location where the plugin has been stored.
+	 * @return array A list of status items.
+	 */
+	public function run_check_plugin_files( $path_to_plugin ) {
+		$this->path_to_plugin = $path_to_plugin;
+		$this->init();
+		$this->prepare_data();
+		$this->run_all_checks();
+		return $this->results;
+	}
+
+	protected function init() {
+
+		// Find all the methods on this class starting with `check_`
+		$methods = get_class_methods( $this );
+		foreach ( $methods as $method ) {
+			if ( 0 === strpos( $method, 'check_' ) )
+				$this->check_methods[] = $method;
+		}
+	}
+
+	protected function prepare_data() {
+		// Parse and stash the readme data
+		$this->readme_path = Import::find_readme_file( $this->path_to_plugin );
+		$this->readme = new Parser( $this->readme_path );
+
+		// Parse and stash plugin headers
+		$this->headers = Import::find_plugin_headers( $this->path_to_plugin );
+
+		// Parse and stash block info
+		$this->blocks = $this->find_blocks( $this->path_to_plugin );
+	}
+
+	public function run_all_checks() {
+		foreach ( array_unique( $this->check_methods ) as $method ) {
+			call_user_func( array( $this, $method ) );
+		}
+	}
+
+	public function find_blocks( $base_dir ) {
```

Comment:
Note that this is a simplified version of the code from class-import.php. It had to be simpler in part because we're operating on a zip file or a single directory and don't have multiple tags available. But there's other logic that certainly belongs here also - it doesn't do as good a job as Corey's improvements to the Import class.